torch/nn/modules/_functions.py

```diff
@@ -0,0 +1,89 @@
+import torch
+from torch.autograd.function import Function
+
+
+class SyncBatchNorm(Function):
+
+    @staticmethod
+    def forward(self, input, weight, bias, running_mean, running_var, eps, momentum, process_group, world_size):
+        input = input.contiguous()
+
+        # calcualte mean/invstd for input.
+        mean, invstd = torch.batch_norm_stats(input, eps)
+
+        mean_all = torch.empty(world_size, mean.size(0), dtype=mean.dtype, device=mean.device)
+        invstd_all = torch.empty(world_size, invstd.size(0), dtype=invstd.dtype, device=invstd.device)
+        mean_l = list(mean_all.unbind(0))
+        invstd_l = list(invstd_all.unbind(0))
+        # using all_gather instead of all reduce so we can calculate mean/var in one go
+        torch.distributed.all_gather(mean_l, mean, process_group)
+        torch.distributed.all_gather(invstd_l, invstd, process_group)
```

Comment:
You can pass async_op=True to run both in parallel instead of serially to improve latency.

Make sure to capture the returned object and call `wait()` on it, before using the output tensors though.

Reply:
Sounds good. Will do.